### PR TITLE
DeNaN pass

### PR DESCRIPTION
**Pull request by kripken**
This moves the fuzzer de-NaN logic out into a separate pass. This is
cleaner and also better since the old way would de-NaN once, but then
the reducer could generate code with nans. The new way lets us de-NaN
while reducing.

### CODE REVIEW

File: src/passes/DeNaN.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      } else {
+        std::cerr << "warning: cannot de-nan outside of function context\n";
+      }
+    }
+  }
+
+  void visitModule(Module* module) {
+    // Add helper functions.
+    Builder builder(*module);
+    auto add = [&](Name name, Type type, Literal literal, BinaryOp op) {
```
Comment (aheejin) [inline src/passes/DeNaN.cpp:65]: Nit: Showing generated code, like this, with explanation that `f32|64.eq` is an ordered comparison that returns false when either operand is a NaN, might help first-time readers (like me :)
```
(if (result f32)
  (f32.eq
    (local.get $0)
    (local.get $0)
  )
  (local.get $0)
  (f32.const 0)
)
```

File: src/passes/pass.cpp [hunk 2]
```diff
@@ -104,6 +104,7 @@ void PassRegistry::registerPasses() {
                createConstHoistingPass);
   registerPass(
     "dce", "removes unreachable code", createDeadCodeEliminationPass);
+  registerPass("denan", "convert all NaNs into 0 at runtime", createDeNaNPass);
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/pass.cpp:107]: The transform is at runtime, but (almost all) checks and conversions are at runtime. I can try to clarify this.

File: src/passes/DeNaN.cpp (hunk 1 above)
Comment (kripken) [reply src/passes/DeNaN.cpp:67]: Good idea, added.

File: src/passes/pass.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/pass.cpp]: Ah right, it calls the function at runtime. I was confused.

**Pull request merged**